### ISSUE TITLE
Support running hooks on all/only tracked files in the repo

### ISSUE THREAD

**Issue by wlritchi** [pre-commit/pre-commit#2818]
### search you tried in the issue tracker

untracked, all_files, all files

### (or, why are you bringing this up again?)

Aspects of the topic here have been discussed a number of times, including in #708, #1212, #1519, #1538, #2194, and #2405. I think the discussions in these feature requests have each been [too focused](https://xyproblem.info/) on one particular resolution that the posters have in mind, and as a result the discussion often doesn't get around to motivating why the posters feel that it's a useful, on-topic feature. Or, by the time it does get around to it, everybody's frustrated, nobody can explain their side very well, and the conversation ends with the maintainer quite reasonably exhausted and the posters feeling unheard.

It's bad form to ask for things that have been asked about before, and also bad form to comment on old requests (which is why they get locked). But that means that if the first person to ask for something does a bad job of arguing their case, it becomes settled law and new points of view can't be brought up. I hope that this issue can collect the arguments for and against the idea more broadly, and also air some of the arguments that I think previous posters haven't managed to articulate very well. Or, at the very least, it's a more comprehensive place to direct people asking about it in the future.

### TL;DR

To the maintainer - you've seen this discussion before, you can jump straight to Objections at the bottom for the arguments that I feel have gone unheard, to hopefully convince you that Possible Solutions is also worth reading.

## Description

Most of the time, pre-commit hooks are best run on only the files that have actually changed, predominantly for speed reasons. However, there are some hooks that users may wish to run on all files. Notably, this may include type checkers, for which a single-file change may introduce problems in other files. For example, removing a property from a type will break anything that tries to access it.

Users currently have a few (unsupported) options to run hooks on all files. For example, some tools like Pyright default to running on the whole repo if given no arguments, and can simply be configured with `pass_filenames: false` like [this community hook](https://github.com/necaris/pre-commit-pyright/blob/master/.pre-commit-hooks.yaml) does. Others can be run via a wrapper or with custom args to pass the repo root (as an example, my employer's smaller repositories generally do this with a mypy wrapper). However, if users run hooks this way, the hooks will also run on any _untracked_ files in the repo, because pre-commit only stashes unstaged changes to tracked files.

I'd like a way to run hooks on all files like this, but only including _tracked_ files.

## A Practical Example

In my workflow, this often shows up when I'm trying to create atomic commits. For example, these commits might have been written together, and now I want to use partial staging to commit them in order:

1. clean up VehicleRegistry
2. add new params to Vehicle.start() interface and impls
3. add Truck implementation

Suppose I add the changes to VehicleRegistry and try to commit. If I have pre-commit running a type checker on my entire repo, it will see the new Truck implementation, because that's a new (untracked) file. However, the changes to the Vehicle interface will have been stashed, so Truck will fail type checking because it doesn't correctly implement the superclass.

### Just stash untracked files 4Head

This has been suggested before in #708 and #1212. This was turned down because some ways of running pre-commit will themselves use untracked files, such as running via tox. Stashing untracked files would delete pre-commit and ruin your day.

I'm not entirely sure that this is as big of a problem as the earlier discussions suggest. `git stash --include-untracked` is also a thing, and it doesn't muck with virtualenvs because it still respects `.gitignore`. I'm not familiar enough with the various pre-commit workflows to evaluate them all, but my gut response is "surely the pre-commit environment is gitignored in any reasonable workflow, right?"

### Allow hooks to receive all files as args, not just changed ones

I think this change would boil down to adding an `all_files` option in the hook config that behaves like a hook-specific version of the `--all-files` command line argument, as suggested in #2194 and #2405 among others. With this option, a mypy hook (for example) could receive every Python file in the repo as its arguments. As far as I can tell, `--all-files` already achieves pretty much the desired behaviour, except for the fact that it affects every hook. This was generally turned down over one of the objections noted below (which I respond to).

### It's the hook's job to do it anyway (i.e. pre-commit Won't Do)

Some might argue that if hooks really want to only run on tracked files, they could check the git status themselves. That's a fair point, but I think there are two reasons why it could make sense for this to be in pre-commit's court:

1. There are multiple hooks that users could reasonably want to run on all tracked files, and really only one reasonable way to implement that check. Just looking at Python type checkers, there are three reasonably well-established ones, and the argument about running on the whole repo applies to all of them. It's not very DRY if the ecosystem of all-files-friendly hooks are all just copying around a wrapper script, and even less DRY if the tool/hook maintainers decide not to do it by default and instead the script is copied around through a thousand how-to blog posts by the users who keep asking how to do it.
2. Most tools run by hooks are not about git, and generally shouldn't care what version control system people are using at all. The hooks themselves generally are relatively thin config or wrappers around the tools, that teach pre-commit how to install and run the tools. pre-commit's job is to take this collection of tools and apply them in a way that's useful with most workflows for one particular revision control system. The change I'm asking for is essentially infrastructure to better represent how users think about "partial commits" in some workflows. From a separation-of-concerns angle, dealing with the consequences of git's implementation of partial commits seems more like pre-commit's job.

### It's the hook's job and somebody should make a community hooks repo for it

Ultimately it's not difficult to write a wrapper to run something on all files, so someone in the community could write this helper script once and create a repo of hooks for the tools that are reasonable to use in an all-files way. This addresses the DRYness problem above, and also serves as at least somewhat of a filter for the community to only recommend (i.e. provide hooks for) running tools this way if it's actually reasonable. So, a hook for pytest would probably be out, but a hook for Pyright might survive.

The downside here is that unless we successfully lobby tool maintainers into upstreaming the wrapper (landing us in the above solution), discoverability of the community hooks repo will probably be low, and users who want this functionality won't find it and will instead keep coming to the pre-commit issue tracker to ask about it anyway.

### Lack of purpose fit

In #2405, you argued that running hooks against all files goes against the idea of the pre-commit framework. Of course, it's any maintainer's right to decide what is and isn't in scope of a project; I'm not blind to the maintenance impact of scope creep or lack of project focus. But I don't think this characterization of pre-commit's purpose is entirely in line with many users' expectations of what pre-commit is about, so I'd like to gently push back. In my mind, the **idea** of pre-commit is in the name: run some checks before committing when they complement the workflow. The **killer feature** is the fact that pre-commit can run only on changed files, because that's compatible with more workflows (including mine) and is much faster, which means less interruption of flow and also more options for hooks that can run in a reasonable time.

In this light, I think it's reasonable to consider that users might find real value in running **some** hooks on all files. Sure, for these particular hooks users miss out on the otherwise-killer speed multiplier, but that's not the only thing that's useful about pre-commit. The killer feature of a Ferrari, I suppose, is that it can go very fast. But that doesn't mean drivers won't sometimes want to drive 5mph because they're in their driveway. Giving users the _option_ to do something a little slow isn't harming them, it's empowering them to decide their own tradeoffs.

I'm not particularly sold on the use case the poster was suggesting in #2405 - pytest really does seem like it's simply too slow to reasonably run in a pre-commit hook. And in #2194, you argued that running hooks against all files is _necessarily_ slow and only rarely catches extra problems. But Pyright runs on my entire (relatively small) repo, or on the pre-commit repo, in a couple of seconds, and has saved me countless time when refactoring because it catches type errors that otherwise _would have_ had to wait for CI to run tests. To me, that's a perfect purpose fit with pre-commit, and I still benefit from all of the other things pre-commit gets right, like convenient setup and isolated tool environments.

And maybe more pointedly: if running on all files is against pre-commit's purpose, then why do we have `pass_filenames: false`?

### Maintenance burden

Any feature inflates the scope of a project, and that needs to be balanced against the amount of code and ongoing maintenance required, with consideration of how many users will actually use it. Again, I'm not ignoring that burden. But I do suspect that the `all_files` approach probably isn't too much of a lift, since `--all-files` does almost exactly the right thing (it's just too broad because it applies to every hook). I also think that, when it's a topic of discussion that keeps coming up (sorry), that's at least an indicator that there are users who want to use it.

## Next steps

Having argued my case, do you feel that this use case is justified? Would you be willing to accept a PR for one of the possible solutions above, or to suggest a community/third-party approach that you'd recommend to the people who still want to run hooks this way?

### pre-commit --version

pre-commit 3.1.1

**Comment by asottile** [comment]
please search the issue tracker, there are several duplicates
